### ISSUE TITLE
Refactoring tool to fix modularization anti-patterns

### ISSUE THREAD

**Issue by mkoeppe** [sagemath/sage#34945]
<div id="comment:0"></div>

The new command `sage --fiximports` replaces imports from `sage.PAC.KAGE.all` by more specific imports when `sage.PAC.KAGE` is an implicit namespace package.

Part of Meta-ticket #34201

Depends on #34849

CC:  @alexchandler100

Component: **scripts**

Author: **Alex Chandler, Matthias Koeppe**

Branch/Commit: **[u/mkoeppe/replace_dot_all](https://github.com/sagemath/sagetrac-mirror/tree/u/mkoeppe/replace_dot_all) @ [`e74dbf6`](https://github.com/sagemath/sagetrac-mirror/commit/e74dbf6c0b9088b3f8b884140bda570de42e0e78)**

Reviewer: **Matthias Koeppe, Alex Chandler**

_Issue created by migration from https://trac.sagemath.org/ticket/34945_

**Comment by mkoeppe** [comment]
Branch: **[u/mkoeppe/replace_dot_all](https://github.com/sagemath/sagetrac-mirror/tree/u/mkoeppe/replace_dot_all)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Commit: **[`294e3e9`](https://github.com/sagemath/sagetrac-mirror/commit/294e3e94a2b510af2212b5a9e11a7c4066cd5ad2)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:2"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/7a546129638dcddce6ef20a7121d39fe240d50da"><code>7a54612</code></a></td><td><code>New command 'sage --fiximports', invokes sage.misc.replace_dot_all</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/294e3e94a2b510af2212b5a9e11a7c4066cd5ad2"><code>294e3e9</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Break some long lines</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`294e3e9`](https://github.com/sagemath/sagetrac-mirror/commit/294e3e94a2b510af2212b5a9e11a7c4066cd5ad2)** to **[`b8b4372`](https://github.com/sagemath/sagetrac-mirror/commit/b8b4372fddfe9b76157fab874ccfacae4a93f739)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:3"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/057dfb5cbf14f61b2ff31bac334febf28d55fa20"><code>057dfb5</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Pass package_regex, not regex including 'import '</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/b8b4372fddfe9b76157fab874ccfacae4a93f739"><code>b8b4372</code></a></td><td><code>src/sage/misc/replace_dot_all.py (make_replacements_in_file): Add optional arg 'output'</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`b8b4372`](https://github.com/sagemath/sagetrac-mirror/commit/b8b4372fddfe9b76157fab874ccfacae4a93f739)** to **[`01efec0`](https://github.com/sagemath/sagetrac-mirror/commit/01efec0040a514ffc005565c505696cd79917ce2)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:4"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/46a15bf3f480669a2e955ff300587c37adc8b351"><code>46a15bf</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Pass package_regex, not regex including 'import ' (fixup)</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/2cbbf2dde134d5836e4e1a3b4b8c18df261c2009"><code>2cbbf2d</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Use more 'with open'</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/29e5ddb4e1a7dd6588bab883d70aed1470bae6fc"><code>29e5ddb</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Replace sort_log_messages by print_log_messages</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/01efec0040a514ffc005565c505696cd79917ce2"><code>01efec0</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Use endswith()</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:5"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/810d9828dfff6d14f258a46d3b7aa42521974e9d"><code>810d982</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Replace -l with positional argument, handle multiple files/dirs, do not use directory prefix</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`01efec0`](https://github.com/sagemath/sagetrac-mirror/commit/01efec0040a514ffc005565c505696cd79917ce2)** to **[`810d982`](https://github.com/sagemath/sagetrac-mirror/commit/810d9828dfff6d14f258a46d3b7aa42521974e9d)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`810d982`](https://github.com/sagemath/sagetrac-mirror/commit/810d9828dfff6d14f258a46d3b7aa42521974e9d)** to **[`dbcccc1`](https://github.com/sagemath/sagetrac-mirror/commit/dbcccc191d38464a75ea5f8b95d124d9ee6159ad)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:6"></div>

Branch pushed to git repo; I updated commit sha1. This was a forced push. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/dbcccc191d38464a75ea5f8b95d124d9ee6159ad"><code>dbcccc1</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Replace -l with positional argument, handle multiple files/dirs, do not use directory prefix</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:7"></div>

Branch pushed to git repo; I updated commit sha1. This was a forced push. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/491805d7a20f231c9194599ccbcd14226a44dc0c"><code>491805d</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Replace -l with positional argument, handle multiple files/dirs, do not use directory prefix</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`dbcccc1`](https://github.com/sagemath/sagetrac-mirror/commit/dbcccc191d38464a75ea5f8b95d124d9ee6159ad)** to **[`491805d`](https://github.com/sagemath/sagetrac-mirror/commit/491805d7a20f231c9194599ccbcd14226a44dc0c)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`491805d`](https://github.com/sagemath/sagetrac-mirror/commit/491805d7a20f231c9194599ccbcd14226a44dc0c)** to **[`bf5e6d2`](https://github.com/sagemath/sagetrac-mirror/commit/bf5e6d22b202818e0a44a945c6c7a629c4aa73fb)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:8"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/8cef4473db1ff3a4a4dc0d308c02fd5da044a3ab"><code>8cef447</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Restore print_log_messages lost in previous commit</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/aa5f1e4ada9ecb2e950186f9ff001fb72ef8c68b"><code>aa5f1e4</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Clean docstrings, uncamelcaps</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/202cdc500232dee3a57d0cf9919ee7827f611c25"><code>202cdc5</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Update doctests</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/bf5e6d22b202818e0a44a945c6c7a629c4aa73fb"><code>bf5e6d2</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Simplify some 'if' conditions</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`bf5e6d2`](https://github.com/sagemath/sagetrac-mirror/commit/bf5e6d22b202818e0a44a945c6c7a629c4aa73fb)** to **[`b426778`](https://github.com/sagemath/sagetrac-mirror/commit/b42677899ad5eee527aa99aaf639a16c7ae22abb)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:9"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/b42677899ad5eee527aa99aaf639a16c7ae22abb"><code>b426778</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Doc updates, pycodestyle fixes</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:10"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/685680bc7f58ee83b937fbb6582aa5c2bd8f21d1"><code>685680b</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Use isinstance instead of type(...) == ...</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`b426778`](https://github.com/sagemath/sagetrac-mirror/commit/b42677899ad5eee527aa99aaf639a16c7ae22abb)** to **[`685680b`](https://github.com/sagemath/sagetrac-mirror/commit/685680bc7f58ee83b937fbb6582aa5c2bd8f21d1)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`685680b`](https://github.com/sagemath/sagetrac-mirror/commit/685680bc7f58ee83b937fbb6582aa5c2bd8f21d1)** to **[`0b8dd2f`](https://github.com/sagemath/sagetrac-mirror/commit/0b8dd2fd78871a280a80eed2c7a5bf08c13c8e58)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:11"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/0b8dd2fd78871a280a80eed2c7a5bf08c13c8e58"><code>0b8dd2f</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Fixup</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`0b8dd2f`](https://github.com/sagemath/sagetrac-mirror/commit/0b8dd2fd78871a280a80eed2c7a5bf08c13c8e58)** to **[`1c19d43`](https://github.com/sagemath/sagetrac-mirror/commit/1c19d434cce937e1d1d322f084eddd23ba240245)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:12"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/d296b7cce10a21fb8f1f706faecd9200ed5f5332"><code>d296b7c</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Replace 'row number', 'line number' by :</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/ac83a49dd9a8dd722df050dbaf32689e0b43dad9"><code>ac83a49</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Use __import__</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/fe617e972b0cf42fd8a1dfba6fc2f783983732c4"><code>fe617e9</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Fold parse_arguments into main</code></td></tr><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/1c19d434cce937e1d1d322f084eddd23ba240245"><code>1c19d43</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Fix doctests</code></td></tr></table>

**Comment by mkoeppe** [comment]
Reviewer: **Matthias Koeppe, ...**

**Comment by mkoeppe** [comment]
Changed author from **Alex Chandler** to **Alex Chandler, Matthias Koeppe**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:14"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/23dff24279aaa285c1ab9d4ad3c5e8b44c0fc1ea"><code>23dff24</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Simplify print_log_messages, fold into main</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`1c19d43`](https://github.com/sagemath/sagetrac-mirror/commit/1c19d434cce937e1d1d322f084eddd23ba240245)** to **[`23dff24`](https://github.com/sagemath/sagetrac-mirror/commit/23dff24279aaa285c1ab9d4ad3c5e8b44c0fc1ea)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`23dff24`](https://github.com/sagemath/sagetrac-mirror/commit/23dff24279aaa285c1ab9d4ad3c5e8b44c0fc1ea)** to **[`568c538`](https://github.com/sagemath/sagetrac-mirror/commit/568c5385bcd4b08bbea0a7975d4c830037caf84d)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:15"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/568c5385bcd4b08bbea0a7975d4c830037caf84d"><code>568c538</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Use os.path.join more</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:16"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/eef9cd5771b986c7cb68597fcf9cfcca2b1b4507"><code>eef9cd5</code></a></td><td><code>src/doc/en/developer/packaging_sage_library.rst: Mention relint, sage --fiximports</code></td></tr></table>

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`568c538`](https://github.com/sagemath/sagetrac-mirror/commit/568c5385bcd4b08bbea0a7975d4c830037caf84d)** to **[`eef9cd5`](https://github.com/sagemath/sagetrac-mirror/commit/eef9cd5771b986c7cb68597fcf9cfcca2b1b4507)**

**Comment by mkoeppe** [comment]
<div id="comment:17" align="right">comment:17</div>

Alex, your code works well. I have only made minor changes that bring the command line interface in line with similar tools, and some style changes.

It has proper doctest coverage now and passes the code style tests.

If you have a chance to look through my changes, you can add your name to "Reviewers"

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
Changed commit from **[`eef9cd5`](https://github.com/sagemath/sagetrac-mirror/commit/eef9cd5771b986c7cb68597fcf9cfcca2b1b4507)** to **[`e74dbf6`](https://github.com/sagemath/sagetrac-mirror/commit/e74dbf6c0b9088b3f8b884140bda570de42e0e78)**

**Comment by 7ed8c4ca-6d56-4ae9-953a-41e42b4ed313** [comment]
<div id="comment:18"></div>

Branch pushed to git repo; I updated commit sha1. New commits:
<table><tr><td><a href="https://github.com/sagemath/sagetrac-mirror/commit/e74dbf6c0b9088b3f8b884140bda570de42e0e78"><code>e74dbf6</code></a></td><td><code>src/sage/misc/replace_dot_all.py: Add copyright notice</code></td></tr></table>

**Comment by mkoeppe** [comment]
Dependencies: **#34849**

**Comment by mkoeppe** [comment]
<div id="comment:20" align="right">comment:20</div>

The Lint and Build&Test failures are unrelated to the changes here.

**Comment by alexchandler100** [comment]
<div id="comment:21" align="right">comment:21</div>

I've taken a look, and the changes all look great to me. I've added my name to the reviewers.

**Comment by alexchandler100** [comment]
Changed reviewer from **Matthias Koeppe, ...** to **Matthias Koeppe, Alex Chandler**

**Comment by mkoeppe** [comment]
Merged in 10.0.beta0
